File: hw4/task2.scala
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

import java.io.{File, PrintWriter}
import scala.collection.mutable
import scala.collection.mutable.{ArrayBuffer, Map, Queue, Set}//, Set => MutableSet}


object task2 {
  def GNA(group: mutable.Map[String, mutable.Set[String]], nodes: Array[String]): Array[((String, String), Double)] = {
    val b_v = mutable.Map.empty[(String, String), Double]

    nodes.foreach { root =>
      val parent = mutable.Map.empty[String, mutable.Set[String]]
      val level = mutable.Map(root -> 0)
      val q = Queue[String](root)
      val visit = mutable.Set(root)
      val n_stp = mutable.Map(root -> 1.0)
      val path = ArrayBuffer(root)

      while (q.nonEmpty) {
        val current = q.dequeue()
        group.getOrElse(current, Set.empty).foreach { nb =>
          if (!visit.contains(nb)) {
            level += nb -> (level(current) + 1)
            visit += nb
            q.enqueue(nb)
            path += nb
            parent.getOrElseUpdate(nb, mutable.Set.empty) += current
            n_stp += nb -> (n_stp.getOrElse(nb, 0.0) + n_stp(current))
          } else if (level(nb) == level(current) + 1) {
            parent(nb) += current
            n_stp(nb) += n_stp(current)
          }
        }
      }

      val n_w = mutable.Map.empty[String, Double].withDefaultValue(1.0)
      val e_w = mutable.Map.empty[(String, String), Double].withDefaultValue(0.0)

      path.reverseIterator.foreach { rv =>
        parent.getOrElse(rv, Set.empty).foreach { b =>
          val w = n_w(rv) * (n_stp(b) / n_stp(rv))
          n_w(b) += w
          val edge_n = if (rv < b) (rv, b) else (b, rv)
          e_w(edge_n) += w
        }
      }

      e_w.foreach { case (key, value) =>
        b_v(key) = b_v.getOrElse(key, 0.0) + value / 2
      }
    }

    implicit val ordering: Ordering[((String, String), Double)] =
      Ordering.by[((String, String), Double), (Double, (String, String))](
        x => (-x._2, x._1)
      )
    b_v.toArray.sorted(ordering)
  }

  def calMod(
              comits: Array[Array[String]], // Assuming communities is an array of array of strings (node IDs)
              edgesG: Map[String, Set[String]], // Edges from a node to its set of neighbors
              m: Int, // Total number of edges
              kN: Map[String, Int] // Degree of each node
            ): Double = {
    var modularity = 0.0
    for (community <- comits) { // Iterate over each community
      for (i <- community) { // Iterate over each node i in the community
        for (j <- community) { // Iterate over each node j in the community
          val aIj = if (edgesG.getOrElse(i, Set.empty).contains(j)) 1.0 else 0.0
          modularity += (aIj - (kN(i) * kN(j)).toDouble / (2.0 * m))
        }
      }
    }
    modularity / (2.0 * m)
  }
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("Task2")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    if (args.length != 4) {
      println("Usage: spark-submit --class Task2 <jar file> <threshold> <input_file> <betweenness_output> <community_output>")
      System.exit(1)
    }

    val threshold = args(0).toInt
    val inputPath = args(1)
    val betweenOutput = args(2)
    val communityOutput = args(3)

    val start = System.currentTimeMillis()

    val rawData = sc.textFile(inputPath)
    val header = rawData.first()
    val data = rawData.filter(_ != header)
      .map(_.split(","))
      .map(parts => (parts(0), Set(parts(1))))
      .reduceByKey(_ union _)

    // Filter users with at least `threshold` products, and prepare data structures for graph construction
    val data_pre = data.filter(_._2.size >= threshold).collectAsMap()
    val uni_user = data_pre.keys.toList
    val combuser = uni_user.combinations(2).toList// All possible pairs of users

    // Filter pairs of users with at least `threshold` shared products, and construct the graph
    val filter_p = sc.parallelize(combuser).filter(users => data_pre(users(0)).intersect(data_pre(users(1))).size >= threshold)
    val edges_n = filter_p.flatMap(users => Seq((users(1), users(0)), (users(0), users(1))))
    val nodes = filter_p.flatMap(users => Seq(users(0), users(1))).distinct()
    val n_list_a: List[String] = nodes.collect().toList

    val edges_kv = edges_n.map { case (user, friend) => (user, Set(friend)) }
    val edges_gr = edges_kv.reduceByKey(_ union _)
    val edges_g_e = edges_gr.collectAsMap()
    val edges_g: mutable.Map[String, mutable.Set[String]] = edges_g_e.map {
      case (key, immutableSet) => key -> (mutable.Set.empty[String] ++= immutableSet)
    }(collection.breakOut)
    val n_list = n_list_a.toArray

    var betweens: Array[((String, String), Double)] = GNA(edges_g, n_list)
    val output_1 = betweens.map { case (pair, between) =>
      (pair, BigDecimal(between).setScale(5, BigDecimal.RoundingMode.HALF_UP).toDouble)
    }
    val pw = new PrintWriter(new File(betweenOutput))
    output_1.map { case ((user1, user2), score) =>
      s"('$user1', '$user2'),$score"
    }.foreach(pw.println)
    pw.close()

    val m = betweens.length
    val l_g: RDD[(String, Int)] = edges_gr.map { case (node, edges) => (node, edges.size) }
    val k_n: mutable.Map[String, Int] = mutable.Map(l_g.collectAsMap().toSeq: _*)
    var max_mod = Double.NegativeInfinity
    var final_com: List[List[String]] = List()
    //var lo = betweens
    while (betweens.nonEmpty) {
      val ((node1, node2), _) = betweens.head
      // removes the selected edge from the graph to potentially fragment the network into communities
      edges_g.get(node1).foreach(_.remove(node2))
      edges_g.get(node2).foreach(_.remove(node1))

      var comits = ArrayBuffer[ArrayBuffer[String]]()
      var oriN = mutable.Set(n_list: _*)

      while (oriN.nonEmpty) {
        val root = oriN.head
        oriN.remove(root)
        val queue = Queue[String](root)
        val visitAg = mutable.Set[String](root)
        val com = ArrayBuffer[String](root)

        while (queue.nonEmpty) {
          val current = queue.dequeue()
          edges_g.getOrElse(current, mutable.Set.empty).foreach { nb =>
            if (!visitAg.contains(nb)) {
              visitAg += nb
              oriN -= nb
              queue.enqueue(nb)
              com += nb
            }
          }
        }
        comits += com.sorted
      }
      val curMod = calMod(comits.map(_.toArray).toArray, edges_g, m, k_n)

      if (curMod > max_mod) {
        max_mod = curMod
        //println(max_mod)
        final_com = comits.map(_.toList).toList
      }

      betweens = GNA(edges_g, n_list) // Recalculate betweenness
    }
    val result = final_com.sortBy(comm => (comm.size, comm.head))
    //result.foreach(println)
    try {
      val writer = new PrintWriter(new File(communityOutput))
      try {
        result.foreach(f => {
          val com = f.map(u => s"'$u'").mkString(", ")
          writer.write(com + "\n")
        })
      } finally {
        writer.close()
      }
    } catch {
      case e: Exception =>
        e.printStackTrace() // Print any exception to understand the failure
    }


    sc.stop()

    val end_time = System.currentTimeMillis()
    val execution = (end_time - start) / 1000.0
    println(s"Duration: $execution")
  }
}

File: hw4/task2.py
import sys
from pyspark import SparkContext
import os
import time
from itertools import combinations
from collections import deque

def GNA(group, nodes):
    b_v = {}
    for root in nodes:
        parent = {}
        level = {}
        q = deque([root])
        # print(queues)
        visit = set()
        visit.add(root)
        level[root] = 0
        n_stp = {root: 1}
        path = [root]
        while q:
            current = q.popleft()
            for nb in group.get(current, []):
                if nb not in visit:#if not visit neighbour
                    level[nb] = level[current] + 1
                    visit.add(nb)
                    q.append(nb)
                    path.append(nb)
                    if nb not in parent.keys():
                        parent[nb] = set()
                    parent[nb].add(current)
                    n_stp[nb] = n_stp.get(nb, 0) + n_stp.get(current, 0)
                elif nb in visit and level[nb] == level[current] + 1:
                    parent[nb].add(current)
                    # refresh the number of shortest paths
                    n_stp[nb] += n_stp[current]

        n_w = {}
        e_w = {}
        rev_path = list(reversed(path))
        for rv in rev_path:
            n_w[rv] = 1
        for a in rev_path:
            for b in parent.get(a, []):
                w = n_w.get(a, 1) * (n_stp.get(b, 1) / n_stp.get(a, 1))
                n_w[b] = n_w.get(b, 1) + w
                edge_n = tuple((a, b)) if a < b else tuple((b, a))

                e_w[edge_n] = e_w.get(edge_n, 0) + w

        for key, value in e_w.items():
            b_v[key] = b_v.get(key, 0) + value / 2

    b = sorted(b_v.items(), key=lambda x: (-x[1], x[0]))
    return b

def cal_mod(c, edge, mx, k_nn):
    modularity = 0.0
    for community in comits:
        for i in community:
            for j in community:
                a_ij = 1 if j in edges_g.get(i, []) else 0.0
                modularity += (a_ij - (k_n[i] * k_n[j]) / (2.0 * m))
    return modularity / (2.0 * m)

if __name__ == "__main__":

    #check the validation of the command
    if len(sys.argv) != 5:
        print("There is an error for the Usage: python3 task2.py <threshold> <input_filepath> <betweenness_output_filepath> <community_output_filepath>")
        sys.exit(1)

    sc = SparkContext(appName="Task2")

    thr = int(sys.argv[1])
    input_file = sys.argv[2]
    between_output = sys.argv[3]
    community_output = sys.argv[4]

    start_time = time.time()
    sc.setLogLevel("ERROR")
    csv_rdd = sc.textFile(input_file)
    header = csv_rdd.first()
    data_nh = (csv_rdd.filter(lambda x: x != header).map(lambda x: x.split(","))
               .map(lambda x: (x[0], {x[1]}))
               .reduceByKey(lambda a, b: a.union(b)))
    data_pre = data_nh.filter(lambda x: len(x[1]) >= thr).collectAsMap()
    uni_user = list(data_pre.keys())
    combuser = list(combinations(uni_user, 2))
    filter_p = sc.parallelize(combuser).filter(
        lambda users: len(data_pre[users[0]].intersection(data_pre[users[1]])) >= thr)
    edges_n = filter_p.flatMap(lambda users: [(users[1], users[0]), (users[0], users[1])])
    nodes = filter_p.flatMap(lambda users: [users[0], users[1]]).distinct().map(lambda user: user)
    n_list = list(nodes.collect())
    edges_kv = edges_n.map(lambda x: (x[0], {x[1]}))
    edges_gr = edges_kv.reduceByKey(lambda a, b: a.union(b))
    edges_g = edges_gr.collectAsMap()
    betweens = GNA(edges_g, n_list)
    #print(betweens)
    output_1 = [(pair, round(between, 5)) for pair, between in betweens]
    with open(between_output, "w") as f:
        for users, value in output_1:
            f.write(f"{users},{value}\n")
    m = len(betweens) #determine m
    l_g = edges_gr.map(lambda x: (x[0], len(x[1]))) #determine the k_i,k_j
    k_n = l_g.collectAsMap() #make as a dictionary

    max_mod = -float('inf')
    final_com = []

    while betweens:
        edge_rm, _ = betweens[0]
        #remove the nodes from the original dictionary
        edges_g[edge_rm[0]].discard(edge_rm[1])
        edges_g[edge_rm[1]].discard(edge_rm[0])

        comits = []
        ori_n = set(n_list)
        while ori_n:
            root = ori_n.pop()
            queue = deque([root])
            visit_ag = set([root])
            com = [root]
            while queue:
                current = queue.popleft()
                for nb in edges_g.get(current, []):
                    if nb not in visit_ag:
                        visit_ag.add(nb)
                        ori_n.discard(nb)
                        queue.append(nb)
                        com.append(nb)
            #struct the splited communities
            comits.append(sorted(com))

        cur_mod = cal_mod(comits, edges_g, m, k_n)

        if cur_mod > max_mod:
            max_mod = cur_mod
            final_com = [list(comm) for comm in comits]  # Avoid deep copy by reconstructing lists

        betweens = GNA(edges_g, n_list)
    result = sorted(final_com, key=lambda x: (len(x), x[0]))
    with open(community_output, "w") as f:
        for use in result:
            com = ', '.join([f"'{u}'" for u in use])
            f.write(com + "\n")

    #---------------------------------------------------------------------------

    end_time = time.time()
    execution = end_time - start_time
    print(f"Duration: {execution}")
    sc.stop()
